zio/ndjsonio/config.go

```diff
@@ -0,0 +1,61 @@
+package ndjsonio
+
+import (
+	"fmt"
+)
+
+// A Rule contains one or more matches and the name of a descriptor
+// key (in the companion Descriptors map).
+type Rule struct {
+	Name       string `json:"name"`
+	Value      string `json:"value"`
+	Descriptor string `json:"descriptor"`
+}
+
+// A TypeConfig contains a map of Descriptors, keyed by name, and a
+// list of rules defining which records should be mapped into which
+// descriptor.
+type TypeConfig struct {
+	Descriptors map[string][]interface{} `json:"descriptors"`
+	Rules       []Rule                   `json:"rules"`
+}
+
+func hasField(name string, columns []interface{}) bool {
+	for _, colmap := range columns {
+		col := colmap.(map[string]interface{})
+		if col["name"] == name {
+			return true
+		}
+	}
+	return false
+}
+
+// Validate validates a typing config.
+func (conf TypeConfig) Validate() error {
+	for _, rule := range conf.Rules {
+		d, ok := conf.Descriptors[rule.Descriptor]
+		if !ok {
+			return fmt.Errorf("rule %s=%s uses descriptor %s that does not exist", rule.Name, rule.Value, rule.Descriptor)
+		}
+		if !hasField(rule.Name, d) {
+			return fmt.Errorf("rule %s refers to field %s that is not present in descriptor", rule.Descriptor, rule.Name)
+		}
+
+	}
+	for name, desc := range conf.Descriptors {
+		col, ok := desc[0].(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("descriptor %s has invalid structure", name)
+		}
+		if col["name"] != "_path" {
+			return fmt.Errorf("descriptor %s does not have _path as first column", name)
+		}
+		for _, d := range desc {
+			col := d.(map[string]interface{})
```

Comment:
Is it possible that this cast fails and panics?

Reply:
oops... fixed and added tests for this